loadgen/ebbandflow/fairness_tracker.go

```diff
@@ -0,0 +1,454 @@
+	JainsFairnessIndex     float64           `json:"jainsFairnessIndex"`
+	CoefficientOfVariation float64           `json:"coefficientOfVariation"`
+	AtkinsonIndex          float64           `json:"atkinsonIndex"`
+	Violations             map[string]string `json:"violations"` // key is fairness indicator
+	TopViolators           []ViolatorReport  `json:"topViolators"`
+}
+
+func NewFairnessTracker() *FairnessTracker {
+	return &FairnessTracker{
+		latencies: make(map[string]*latencyHeap),
+		weights:   make(map[string]float32),
+	}
+}
+
+func (ft *FairnessTracker) Track(
+	fairnessKey string,
+	fairnessWeight float32,
+	scheduleToStartLatency time.Duration,
+) {
+	if fairnessKey == "" {
+		return
+	}
+	ft.mu.Lock()
+	defer ft.mu.Unlock()
+
+	h, exists := ft.latencies[fairnessKey]
+	if !exists {
+		h = &latencyHeap{}
+		heap.Init(h)
+		ft.latencies[fairnessKey] = h
+	}
+	heap.Push(h, float64(scheduleToStartLatency.Milliseconds()))
+	ft.weights[fairnessKey] = fairnessWeight
+}
+
+// GetReport returns a processed fairness report without clearing the data.
+func (ft *FairnessTracker) GetReport() (*FairnessReport, error) {
+	ft.mu.RLock()
+	if len(ft.latencies) < 2 {
+		ft.mu.RUnlock()
+		return nil, fmt.Errorf("need at least 2 fairness keys, got %d", len(ft.latencies))
+	}
+
+	heapData := make(map[string][]float64, len(ft.latencies))
+	weights := make(map[string]float64, len(ft.weights))
+	for key, h := range ft.latencies {
+		if h.Len() == 0 {
+			continue
+		}
+		values := make([]float64, h.Len())
+		copy(values, *h)
+		heapData[key] = values
+
+		// Get the weight for this key
+		if weight, exists := ft.weights[key]; exists {
+			weights[key] = float64(weight)
+		} else {
+			weights[key] = 1.0 // default weight
+		}
+	}
+	ft.mu.RUnlock()
+
+	p95Values := make(map[string]float64, len(heapData))
+	sampleCounts := make(map[string]int, len(heapData))
+	for key, values := range heapData {
+		sort.Float64s(values)
+		p95 := calculatePercentile(values, 0.95)
+		p95Values[key] = p95
+		sampleCounts[key] = len(values)
+	}
+
+	if len(p95Values) < 2 {
+		return nil, fmt.Errorf("need at least 2 keys with data, got %d", len(p95Values))
+	}
+
+	// Extract P95 values for distribution analysis
+	p95Slice := make([]float64, len(p95Values))
+	i := 0
+	for _, p95 := range p95Values {
+		p95Slice[i] = p95
+		i++
+	}
+	sort.Float64s(p95Slice)
+
+	// Calculate weight-adjusted P95 values for all metrics.
+	weightAdjustedP95s := make(map[string]float64, len(p95Values))
+	for key, p95 := range p95Values {
+		if weight := weights[key]; weight > 0 {
+			weightAdjustedP95s[key] = p95 / weight
+		}
+	}
+
+	// Approach 1: Weight-Adjusted Proportional Fairness
+	// Normalizes each key's P95 by weight (P95/weight), then checks if the distribution
+	// of normalized values is flat. Higher weight keys should get lower latencies, so
+	// after normalization all keys should have similar values. We compare P95-of-normalized
```

Comment:
yes, keys with higher FairnessWeight should have lower latencies, but we're dividing by FairnessWeight above, right? so a lower number divided by a higher number is doubly lower... we should multiply instead